Markdown 128-139 (#59)
ultimate-go-korean/translation: 128-139/128-139.md

#### Data race

##### Race Detection

> As soon as we add another Goroutine to our program, we add a huge amount of complexity. We can't always let the Goroutine run stateless. There has to be coordination. There are, in fact, 2 things that we can do with multi threaded software. - We either have to syncronize access to share state like that WaitGroup is done with Add, Done and Wait. - Or we have to coordinate these Goroutines to behave in a predictable or responsible manner.

Reviewer: 원문에서는 리스트로 안나누어져있나요?
그냥 바로 리스트로 빼는게 보기 더 좋아보입니다.

```suggestion
As soon as we add another Goroutine to our program, we add a huge amount of complexity. We can't always let the Goroutine run stateless. There has to be coordination. There are, in fact, 2 things that we can do with multi-threaded software. 

- We either have to synchronize access to share state like that WaitGroup is done with `Add`, `Done` and `Wait`. 
- Or we have to coordinate these Goroutines to behave in a predictable or responsible manner.
```

Author: @kkweon 네 이부분은 수정한게 반영이 안되고올라갔나보네요ㅠㅠ 수정했습니다

Up until the use of channel, we have to use atomic function, mutex, to do both. The channel gives us a simple way to do orchestration. However, in many cases, using atomic function, mutex, and synchronizing access to shared state is the best way to go.

Atomic instructions are the fastest way to go because deep down in memory, Go is syncronizing 4-8 bytes at a time.

Mutexes are the next fastest. Channels are very slow because not only they are mutexes, there are all data structures and logic that go with them. Data races happen when we have multiple Goroutines trying to access the same memory location. For example, in the simplest case, we have an integer that is a counter. We have 2 Goroutines that want to read and write to that variable at the same time. If they are actually doing it at the same time, they are going to trash each other read and write. Therefore, this type of synchronizing access to the shared state has to be coordinated.

> The problem with data races is that they always appear random. Sample program to show ho to create race conditions in our programs. We don't want to do this.

Reviewer: ```suggestion
The problem with data races is that they always appear random. Sample program to show how to create race conditions in our programs. We don't want to do this.
```

Reviewer 2: @kkweon 님 이걸 어떻게 찾아내셨데요. @@

```go
package main

import (
    "fmt"
    "runtime"
    "sync"
)
```

counter is a variable incremented by all Goroutines.

```go
var counter int

func main() {
```

Number of Goroutines to use.

```go
    const grs = 2
```

wg is used to manage concurrency.

```go
    var wg sync.WaitGroup
    wg.Add(grs)
```

Create two Goroutines.

They loop twice: perform a read to a local counter, increase by 1, write it back to the shared state. Every time we run the program, the output should be 4.

> The data races that we have here is that: at any given time, both Goroutines could be reading and writing at the same time. However, we are very lucky in this case. What we are seeing is that each Goroutine is executing the 3 statements atomically comletely by accident every time this code run.

Reviewer: ```suggestion
The data races that we have here is that: at any given time, both Goroutines could be reading and writing at the same time. However, we are very lucky in this case. What we are seeing is that each Goroutine is executing the 3 statements atomically completely by accident every time this code run.
```

Previous write at 0x000001228340 by goroutine 7: main.main.func1()
/Users/hoanhan/work/hoanhan101/ultimate-go/go/concurrency/data_race_1.go

:75 +0x68
Goroutine 8 (running) created at: main.main()
/Users/hoanhan/work/hoanhan101/ultimate-go/go/concurrency/data_race_1.go :62 +0xab
Goroutine 7 (finished) created at: main.main()
/Users/hoanhan/work/hoanhan101/ultimate-go/go/concurrency/data_race_1.go :62 +0xab
==================
Final Counter: 4
Found 1 data race(s)
exit status 66
```

##### Atomic Functions

```go
package main

import (
    "fmt"
    "runtime"
    "sync"
    "sync/atomic"
)
```

counter is a variable incremented by all Goroutine. Notice that it's not just an int but int64. We are being very specific about the precision because the atomic function requires us to do so.

```go
var counter int64

func main() {
```

Number of Goroutines to use.

```go
    const grs = 2
```

wg is used to manage concurrency.

```go
    var wg sync.WaitGroup
    wg.Add(grs)
```

Create two goroutines.

```go
    for i := 0; i < grs; i++ {
        go func() {
            for count := 0; count < 2; count++ {
```

> Safely add one to the counter. Add the atomic functions that we have taken an address as the first parameter and that is being syncronized, no matter how many Goroutines they are. If we call one of these functions on the same location, they will get serialized. This the fastest way to serialization.

Reviewer: ```suggestion
Safely add one to the counter. Add the atomic functions that we have taken an address as the first parameter and that is being syncronized, no matter how many Goroutines they are. If we call one of these functions on the same location, they will get serialized. This the fastest way to serialize.
```

Author: @kkweon 네 이부분은 This is the fastest way to serialization 이 맞더라구요ㅠㅠ 그래서 해당 부분은 본문 보고 수정했습니다.

We can run thist program all day long and still get 4 every time.

```go
                atomic.AddInt64(&counter, 1)
```

This call is now irrelevant because by the time AddInt64 function completes, counter is increment.

```go
                runtime.Gosched()
            }

        wg.Done()
        }()
    }
```

Wait for the Goroutines to finish.

```go
    wg.Wait()
```

Display the final value.

```go
    fmt.Println("Final Counter:", counter)
}
```

```shell
Final Counter: 4
```

##### Mutexes

We don't always have the luxury of using 4-8 bytes of memory as shared data. This is where the mutex comes in. Mutex allows us to have an API like the WaitGroup (Add, Done and Wait) where any Goroutine can execute one at a time.

```go
package main

import (
    "fmt"
    "sync"
)

var (
```

counter is a variable increment by all Goroutines.

```go
    counter int
```

> mutex is used define a critical section of code. Picture mutex as a room where all Goroutines have to go through. However, only one Goroutine can go at a time. The scheduler will decide who can get in and which one is next. We cannot determine what the scheduler is gonna do. Hopefully, it is gonna be fair. Just because one Goroutine got to the door before another, it doesn't mean that Goroutine will get to the end first. Nothing here is predictable.

Reviewer: ```suggestion
mutex is used to define a critical section of code. Picture mutex as a room where all Goroutines have to go through. However, only one Goroutine can go at a time. The scheduler will decide who can get in and which one is next. We cannot determine what the scheduler is gonna do. Hopefully, it is gonna be fair. Just because one Goroutine got to the door before another, it doesn't mean that Goroutine will get to the end first. Nothing here is predictable.
```

The key here is, once a Goroutine is allowed in, it must report that it's out. All the Goroutines will ask for a lock and unlock when they leave for another one to get in. Two different functions can use the same mutex which means only one Goroutine can execute any of given functions at a time.

```go
    mutex sync.Mutex
)
```

```go
func main() {
```

Number of Goroutines to use.

```go
    const grs = 2
```

wg is used to manage concurrency.

```go
    var wg sync.WaitGroup
    wg.Add(grs)
```

Create two Goroutines.

```go
    for i := 0; i < grs; i++ {
        go func() {
            for count := 0; count < 2; count++ {
```

> Only allow one Goroutine through this critical section at a time. Creating these artifical curly brakets gives readability. We don't have to do this but it is highly recommended. The Lock and Unlock function must always be together in line of sight.

Reviewer: ```suggestion
Only allow one Goroutine through this critical section at a time. Creating these artificial curly brackets gives readability. We don't have to do this but it is highly recommended. The Lock and Unlock function must always be together in line of sight.
```

```go
                mutex.Lock()
                {
```

Capture the value of counter.

```go
                    value := counter
```

Increment our local value of counter.

```go
                    value++
```

Store the value back into counter.

```go
                    counter = value
                }
                mutex.Unlock()
```

Relese the lock and allow any waiting Goroutine through.

```go
            }

            wg.Done()
        }()
    }
```

Wait for the Goroutines to finish.

```go
    wg.Wait()
    fmt.Printf("Final Counter: %d\n", counter)
}
```

```shell
Final Counter: 4
```

##### Read/Write Mutex

> There are times when we have a shared resources where we want many Goroutines reading it.

Reviewer: ```suggestion
There are times when we have shared resources where we want many Goroutines reading it.
```

Occasionally, one Goroutine can come in and make change to the resource. When that happens, everybody has to stop reading. It doesn't make sense to synchronize reads in this type of scenario because we are just adding latency to our software for no reason.

```go
package main

import (
    "fmt"
    "math/rand"
    "sync"
    "sync/atomic"
    "time"
)
```

data is a slice that will be shared.

```go
var (
    data []string
```

> rwMutex is used to define a critical section of code. It is a little bit slower than Mutex but we are optimizing the correctness first so we don't care about that for now.

Reviewer: ```suggestion
`rwMutex` is used to define a critical section of code. It is a little bit slower than Mutex but we are optimizing the correctness first so we don't care about that for now.
```

```go
    rwMutex sync.RWMutex
```

> Number of reads occuring at any given time. As soon as we see int64 here, we should start thinking about using atomic instruction.

Reviewer: ```suggestion
The number of reads occurring at any given time. As soon as we see `int64` here, we should start thinking about using atomic instruction.
```

Author: @kkweon 해당 부분은 본문을 확인했을때도 135page에서 Number of reads로 표기되어있어서 별도 수정은 하지 않았습니다 나머지 occurring, `int64`는 수정하였습니다 :)

```go
    readCount int64
)
```

> init is called prior to main.

Reviewer: ```suggestion
`init` is called prior to `main`.
```
